### PR TITLE
`gear-node-loader` clean-up

### PR DESCRIPTION
**Pull request by SabaunT**
1. Log activity using specialized crate (tokio's `tracing`). It much simplifies our logging logic.
2. Fix investigating node crash (with event and timeouts).
3. Enable login with Suri in `gear-node-loader`
4. General clean-up
5. Also resolves this #1659
6. Introduce proper nonce management that doesn't block node loading

### CODE REVIEW

File: utils/node-loader/src/batch_pool.rs [hunk 1]
```diff
@@ -26,6 +26,17 @@ mod report;
 
 type Seed = u64;
 
+/*
+1. GearApiProducer (создает новое гир апи из старого, при этом апдейтит нонсы)
+2. Лоадер - Репортер:
+2.1. При инициализации пишет в лог файл с каким сидом начала работать.
```
Comment (grishasobol) [inline utils/node-loader/src/batch_pool.rs:32]: english?)

File: gclient/src/api/listener/mod.rs [hunk 2]
```diff
@@ -263,4 +263,9 @@ pub trait EventProcessor {
             .flatten()
             .collect())
     }
+
+    async fn queue_processing_reverted(&mut self) -> Result<()> {
```
Comment (breathx) [inline gclient/src/api/listener/mod.rs:267]: why not just boolean one?

Comment (breathx) [reply gclient/src/api/listener/mod.rs:267]: cause error can occur on other levels

File: utils/node-loader/src/batch_pool.rs [hunk 3]
```diff
@@ -1,373 +1,255 @@
...
+
+    loop {
+        let r = match api.events_since(block_hash, wait_for_events_blocks).await {
+            Ok(mut v) => v.err_or_succeed_batch(messages.keys().copied()).await,
+            Err(e) => Err(e),
+        };
+
+        // If one block is considered to be produced in 1 second, than we wait for 10000 millis (1 sec) * `wait_for_events_blocks`
+        // We also multiply it on the 5 just to be 100% sure if no events occurred, than node is crashed
+        if (utils::now() - now) as usize > wait_for_events_blocks * 1000 * 5 {
```
Comment (breathx) [inline utils/node-loader/src/batch_pool.rs:206]: block time should be caught from the runtime

File: utils/node-loader/src/batch_pool.rs [hunk 4]
```diff
@@ -1,373 +1,255 @@
...
+            break;
+        }
+    }
 
-            let results = results?;
+    let mut program_ids = if collect_programs {
+        Some(BTreeSet::new())
+    } else {
+        None
+    };
```
Comment (breathx) [inline utils/node-loader/src/batch_pool.rs:223]:
```suggestion
    let mut program_ids = collect_programs.then(BTreeSet::new);
```

File: utils/node-loader/src/batch_pool/batch/create_program.rs [hunk 5]
```diff
@@ -24,21 +26,22 @@ impl CreateProgramArgs {
         let mut rng = Rng::seed_from_u64(rng_seed);
 
         let code_idx = rng.next_u64() as usize;
-        let code = existing_codes
-            .ring_get(code_idx)
-            .copied()
-            .expect("Infallible");
+        let &code = existing_codes.ring_get(code_idx);
```
Comment (breathx) [inline utils/node-loader/src/batch_pool/batch/create_program.rs:29]: strange syntax, but ok

File: utils/node-loader/src/batch_pool/batch/send_message.rs [hunk 6]
```diff
@@ -2,11 +2,14 @@
...
-    utils::{LoaderRng, NonEmptyVec, RingGet},
+    utils::{LoaderRng, NonEmptyVec},
 };
-use gear_core::ids::ProgramId;
+use gclient::Result as GClientResult;
+use gear_core::ids::{MessageId, ProgramId};
+use primitive_types::H256;
 
 pub type SendMessageArgsInner = (ProgramId, Vec<u8>, u64, u128);
+pub type SendMessageBatchOutput = (Vec<GClientResult<(MessageId, ProgramId)>>, H256);
```
Comment (breathx) [inline utils/node-loader/src/batch_pool/batch/send_message.rs:12]: why don't you satisfied with just `Result` without any name mappings?

File: utils/node-loader/src/batch_pool/batch/upload_program.rs [hunk 7]
```diff
@@ -27,10 +31,14 @@ impl UploadProgramArgs {
...
         rng.fill_bytes(&mut payload);
 
-        // TODO: add this.
+        tracing::debug!(
+            "Generated `upload_program` batch with code seed = {code_seed}, salt = {} payload = {}",
+            hex::encode(&salt),
+            hex::encode(&payload)
+        );
+
         let gas_limit = 240_000_000_000;
```
Comment (breathx) [inline utils/node-loader/src/batch_pool/batch/upload_program.rs:40]: for all gas limits also caught it from the runtime please (see gclient/tests how to do it)

**Review by breathx — APPROVED** [review]
lgtm

File: gclient/src/api/listener/mod.rs (hunk 2 above)
Comment (SabaunT) [reply gclient/src/api/listener/mod.rs:267]: Occurred errors do not give 100% confidence about queue processing being stopped. For example, an error we get while processing events stream is `jsonrpsee::Error::ParseError`, which tells nothing about node's queue.

File: utils/node-loader/src/batch_pool.rs (hunk 4 above)
Comment (SabaunT) [reply utils/node-loader/src/batch_pool.rs:223]: done

File: utils/node-loader/src/batch_pool/batch/create_program.rs (hunk 5 above)
Comment (SabaunT) [reply utils/node-loader/src/batch_pool/batch/create_program.rs:29]: resolve?

File: utils/node-loader/src/batch_pool.rs (hunk 3 above)
Comment (SabaunT) [reply utils/node-loader/src/batch_pool.rs]: done

File: gclient/src/api/listener/mod.rs (hunk 2 above)
Comment (breathx) [reply gclient/src/api/listener/mod.rs:267]: Yeah, sorry for my explanation, I'm about returning `Result<bool>`

File: utils/node-loader/src/batch_pool/batch/send_message.rs (hunk 6 above)
Comment (breathx) [reply utils/node-loader/src/batch_pool/batch/send_message.rs:12]: `GClientResult` -> `Result`

Comment (SabaunT) [reply utils/node-loader/src/batch_pool/batch/send_message.rs]: done

File: utils/node-loader/src/batch_pool.rs [hunk 8]
```diff
@@ -45,329 +47,206 @@ impl<Rng: LoaderRng> BatchPool<Rng> {
...
+    let results: GClientResult<Vec<(MessageId, Option<String>)>>;
+    let now = utils::now();
+    // States what amount of blocks we should wait for taking all the events about successful `messages` execution
+    let wait_for_events_blocks = 10;
+    // Multiply on five to be 100% sure if no events occurred, than node is crashed
+    let wait_for_events_millisec = api.expected_block_time()? as usize * wait_for_events_blocks * 5;
+
+    loop {
+        let r = match api.events_since(block_hash, wait_for_events_blocks).await {
+            Ok(mut v) => v.err_or_succeed_batch(messages.keys().copied()).await,
```
Comment (shamilsan) [inline utils/node-loader/src/batch_pool.rs:203]: Calling `messages.keys` in the loop seems not good unless `messages` changes inside the loop.

File: utils/node-loader/src/batch_pool.rs [hunk 9]
```diff
@@ -45,329 +47,206 @@ impl<Rng: LoaderRng> BatchPool<Rng> {
...
-            loop {
-                let r = match api.events_since(batch_block_hash, 10).await {
-                    Ok(mut v) => v.err_or_succeed_batch(init_messages.keys().cloned()).await,
-                    Err(e) => Err(e),
-                };
+        if matches!(r, Err(GClientError::EventNotFoundInIterator)) {
+            continue;
+        } else {
+            results = r;
+            break;
```
Comment (shamilsan) [inline utils/node-loader/src/batch_pool.rs:216]: You may return the result with the break.
```suggestion
            break r;
```

```rust
let a = loop {
    break 42;
};
    
println!("{a}"); // 42
```

File: utils/node-loader/src/utils.rs [hunk 10]
```diff
@@ -41,18 +53,33 @@ pub fn str_to_wsaddr(endpoint: String) -> WSAddress {
...
+        }
+
+        size += 1;
+
+        Some(args())
+    })
+}
+
+pub fn convert_iter<V, T: Into<V>>(args: Vec<T>) -> impl IntoIterator<Item = V> {
+    args.into_iter().map(|t| t.into())
```
Comment (shamilsan) [inline utils/node-loader/src/utils.rs:73]: A few more func-style 😄
```suggestion
    args.into_iter().map(Into::into)
```

Comment (SabaunT) [reply utils/node-loader/src/utils.rs]: done in [8c58df2](https://github.com/gear-tech/gear/pull/1600/commits/8c58df269e07992ad196e1793a6fab68f63d0671)

File: utils/node-loader/src/batch_pool.rs [hunk 11]
```diff
@@ -45,329 +47,206 @@ impl<Rng: LoaderRng> BatchPool<Rng> {
...
 
-            let results = results?;
+        if let Some(expl) = maybe_err {
+            tracing::debug!("[Call with id: {call_id}]: {mid:#.2} executing within program '{pid:#.2}' ended with a trap: '{expl}'");
+        } else {
+            tracing::debug!("[Call with id: {call_id}]: {mid:#.2} successfully executed within program '{pid:#.2}'");
+            program_ids = program_ids.map(|mut ids| {
+                ids.insert(pid);
+                ids
+            });
```
[the remark this answers is not in the document]
Comment (SabaunT) [reply utils/node-loader/src/batch_pool.rs]: done in [8c58df2](https://github.com/gear-tech/gear/pull/1600/commits/8c58df269e07992ad196e1793a6fab68f63d0671)

File: utils/node-loader/src/batch_pool.rs (hunk 9 above)
Comment (SabaunT) [reply utils/node-loader/src/batch_pool.rs]: done in [8c58df2](https://github.com/gear-tech/gear/pull/1600/commits/8c58df269e07992ad196e1793a6fab68f63d0671)

File: utils/node-loader/src/batch_pool.rs (hunk 8 above)
Comment (SabaunT) [reply utils/node-loader/src/batch_pool.rs:202]: discussed

**Review by shamilsan — APPROVED** [review]
Good catch of sync functions under async mask 👍

File: gclient/src/api/listener/mod.rs (hunk 2 above)
Comment (SabaunT) [reply gclient/src/api/listener/mod.rs:267]: discussed

**Review by grishasobol — APPROVED** [review]

**Pull request merged**